### PR TITLE
ftjam: fix apple clang 12 and test package if 2 profiles + modernize

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **lib/1.0**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
even by fixing config.guess and config.sub, cross-build to armv8 won't work because build scripts have to call an executable built for host machine.

**Review by SSE4 — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

File: recipes/ftjam/all/conanfile.py [hunk 1]
```diff
@@ -20,29 +22,39 @@ def _source_subfolder(self):
...
         del self.settings.compiler.libcxx
         del self.settings.compiler.cppstd
 
-    def source(self):
-        tools.get(**self.conan_data["sources"][self.version])
-        os.rename("ftjam-{}".format(self.version), self._source_subfolder)
+    def validate(self):
+        if self.settings.compiler == "Visual Studio":
+            raise ConanInvalidConfiguration("ftjam doesn't build with Visual Studio yet")
+        if hasattr(self, "settings_build") and tools.cross_building(self):
```
Comment (madebr) [inline recipes/ftjam/all/conanfile.py:31]: Isn't `tools.cross_building(self)` sufficient?

File: recipes/ftjam/all/test_package/conanfile.py [hunk 2]
```diff
@@ -6,22 +6,26 @@
 class TestPackage(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
 
+    def build_requirements(self):
+        if hasattr(self, "settings_build"):
+            self.build_requires(str(self.requires["ftjam"]))
```
Comment (madebr) [inline recipes/ftjam/all/test_package/conanfile.py:11]: This causes a bootstrap problem on c3i when adding a new recipe.
I'm not sure what version of the build requirement is going to be used.

See https://github.com/conan-io/conan-center-index/pull/6677#issuecomment-893434563

File: recipes/ftjam/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/ftjam/all/conanfile.py:31]: No, with 1 profile, `tools.cross_building(self)` is True on Linux machine evaluating package id when host is not Linux.

File: recipes/ftjam/all/test_package/conanfile.py (hunk 2 above)
Comment (SpaceIm) [reply recipes/ftjam/all/test_package/conanfile.py:11]: Well, the issue here is that we don't want to test ftjam in host context, but in build context. This trick works as soon as `foo` in `self.build_requires(str(self.requires["foo"]))` is the tested recipe, and not a dependency of this recipe.

Comment (madebr) [reply recipes/ftjam/all/test_package/conanfile.py:11]: So we're fine that we're building `ftjam` for the host machine, but are actually testing `ftjam` for the build machine?

Comment (SpaceIm) [reply recipes/ftjam/all/test_package/conanfile.py:11]: Ahah yes, but cross-build of ftjam doesn't work anyway. Actually here I was able to see this issue while using 2 same profiles.

File: recipes/ftjam/all/conanfile.py (hunk 1 above)
Comment (madebr) [reply recipes/ftjam/all/conanfile.py:31]: Thanks.
Indeed, with 1 profile approach, it falls back on a comparison with `detected_os()` and `detected_architecture()`.

**Review by madebr — APPROVED** [review]

**Pull request merged**